kubeadm: support certificate chain validation

robbiemcmichael opened the pull request:
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

Fixes an issue where some kubeadm phases fail if a certificate file contains a certificate chain with one or more intermediate CA certificates. This two-tier CA configuration is required to allow a short expiry to be used on the CA used to sign certificates without resorting to frequent manual rotations of the root CA.

The proposal is to change the validation algorithm from requiring that a certificate was signed directly by the root CA to requiring that there is a valid certificate chain back to the root CA. The fix itself is quite simple, most of the complexity is in the PKI helper functions required to unit test this change properly.

**Which issue(s) this PR fixes**:

Fixes the first part of https://github.com/kubernetes/kubeadm/issues/2360

**Special notes for your reviewer**:

N/A

**Does this PR introduce a user-facing change?**:

```release-note
kubeadm: add support for certificate chain validation. When using kubeadm in external CA mode, this allows an intermediate CA to be used to sign the certificates. The intermediate CA certificate must be appended to each signed certificate for this to work correctly.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

N/A

neolit123 commented on cmd/kubeadm/app/util/pkiutil/pki_helpers.go:
```diff
@@ -256,6 +292,21 @@ func TryLoadCertFromDisk(pkiPath, name string) (*x509.Certificate, error) {
+	}
+
+	cert := certs[0]
+	intermediates := certs[1:]
```
AFAIK, orders in chains are not guaranteed, thus this feels applicable to your use case only.

neolit123 replied:
if a chain contains no intermediates, this line will panic.

robbiemcmichael replied:
> AFAIK, orders in chains are not guaranteed, thus this feels applicable to your use case only.

I can't find anything explicit in [RFC 5280](https://tools.ietf.org/html/rfc5280) or [RFC 7468](https://tools.ietf.org/html/rfc7468) to say that the leaf certificate must be first. However, I can see that the `crypto/tls` library makes this assumption in its [certificate verification logic](https://github.com/golang/go/blob/663cd862edf8dfa9c78d9df0f372c9bea03139e4/src/crypto/tls/handshake_client.go#L851) as well an explicit comment elsewhere that the [leaf certificate should be first](https://github.com/golang/go/blob/663cd862edf8dfa9c78d9df0f372c9bea03139e4/src/crypto/tls/common.go#L1323).

Since I can't find an explicit statement in the standards I'm going to have to fall back on the argument that if it's good enough for Golang's standard TLS implementation then it's good enough for the certificate verification helpers in kubeadm. The Kubernetes components will be using the TLS implementation anyway, so it makes sense that the kubeadm phases which perform certificate verification behave the same way.

robbiemcmichael replied:
> if a chain contains no intermediates, this line will panic.

It shouldn't panic, if there's only a single certificate then `intermediates` will be an empty slice. Here's a quick example:

```go
package main

import "fmt"

func main() {
    x := []int{1}
    fmt.Printf("Head: %d\n", x[0])
    fmt.Printf("Tail: %v\n", x[1:])
}
```

This produces:

```
Head: 1
Tail: []
```

randomvariable replied:
[RFC5286 7.4.2](https://tools.ietf.org/html/rfc5246) describes the chain order:

> The sender's certificate MUST come first in the list.  Each following certificate MUST directly certify the one preceding it.  Because certificate validation requires that root keys be distributed independently, the self-signed certificate that specifies the root certificate authority MAY be omitted from the chain, under the assumption that the remote end must already possess it in order to validate it in any case.

neolit123 replied:
> It shouldn't panic,

my mistake.

> RFC5286 7.4.2 describes the chain order:

thanks for the info!

neolit123 commented on cmd/kubeadm/app/util/pkiutil/pki_helpers.go:
```diff
@@ -556,32 +625,39 @@ func NewPrivateKey(keyType x509.PublicKeyAlgorithm) (crypto.Signer, error) {
 
 // NewSignedCert creates a signed certificate using the given CA certificate and key
-func NewSignedCert(cfg *CertConfig, key crypto.Signer, caCert *x509.Certificate, caKey crypto.Signer) (*x509.Certificate, error) {
+func NewSignedCert(isCA bool, cfg *CertConfig, key crypto.Signer, caCert *x509.Certificate, caKey crypto.Signer) (*x509.Certificate, error) {
```
this bool should be the last arg ideally.

robbiemcmichael replied:
Change made in this commit: https://github.com/kubernetes/kubernetes/pull/97266/commits/6f90532f2db098c52a531bb17e6db39d23a8db41

neolit123 commented on cmd/kubeadm/app/util/pkiutil/pki_helpers_test.go:
```diff
@@ -412,6 +433,103 @@ func TestTryLoadCertFromDisk(t *testing.T) {
+		t.Errorf(
+			"failed to create intermediate CA cert and key with an error: %v",
+			err,
+		)
```
this t.Errorf does not need to span on multiple lines.

robbiemcmichael replied:
All requested formatting changes have been made in this commit: https://github.com/kubernetes/kubernetes/pull/97266/commits/d43e52497c602089b6d9f9809ca084f0625c695f

neolit123 commented on cmd/kubeadm/app/util/pkiutil/pki_helpers_test.go:
```diff
@@ -412,6 +433,103 @@ func TestTryLoadCertFromDisk(t *testing.T) {
+		t.Errorf(
+			"failed to create leaf cert and key with an error: %v",
+			err,
+		)
```
ditto

neolit123 commented on cmd/kubeadm/app/util/pkiutil/pki_helpers_test.go:
```diff
@@ -412,6 +433,103 @@ func TestTryLoadCertFromDisk(t *testing.T) {
+		t.Errorf(
+			"failed to write cert: %v",
+			err,
+		)
```
ditto

neolit123 commented on cmd/kubeadm/app/util/pkiutil/pki_helpers_test.go:
```diff
@@ -412,6 +433,103 @@ func TestTryLoadCertFromDisk(t *testing.T) {
+	if err != nil {
+		t.Errorf(
+			"failed to write cert bundle: %v",
+			err,
```
ditto

neolit123 commented on cmd/kubeadm/app/phases/certs/certs.go:
```diff
@@ -416,20 +422,28 @@ func validateSignedCert(l certKeyLocation) error {
+	if err := pkiutil.VerifyCertChain(signedCert, intermediates, caCert); err != nil {
 		return errors.Wrapf(err, "certificate %s is not signed by corresponding CA", l.uxName)
 	}
+
```
we should exclude whitespace changes from the PR.

neolit123 commented on cmd/kubeadm/app/util/pkiutil/pki_helpers.go:
```diff
@@ -624,3 +700,27 @@ func ValidateCertPeriod(cert *x509.Certificate, offset time.Duration) error {
+
+	if _, err := cert.Verify(verifyOptions); err != nil {
+		return err
+	}
```
this change doesn't seem backwards compatible to clusters created with older versions of kubeadm.
calling VerifyCertChain would now result in an error if the intermediate pool is empty, correct?

robbiemcmichael replied:
It is backwards compatible since the `Verify` function doesn't require the intermediate pool to contain any certificates, it only needs to contain the certificates required to construct a valid chain. No additional certificates are required when the leaf certificate is signed directly by the root CA, so this function will succeed.

I've also only changed the verification logic and haven't changed anything to do with the way that kubeadm generates certificates, so the standard setup will still be where the root CA directly signs the leaf certificate.

robbiemcmichael replied:
I've added an extra test case to prove that this is backwards compatible: https://github.com/kubernetes/kubernetes/pull/97266/commits/511f498491abd3786644044138364e04e38dd536

randomvariable replied:
Yup, this should work fine AFAIK.

neolit123 commented on cmd/kubeadm/app/util/pkiutil/pki_helpers.go:
```diff
@@ -79,14 +80,29 @@ func NewCertificateAuthority(config *CertConfig) (*x509.Certificate, crypto.Sign
 }
 
+// NewIntermediateCertificateAuthority creates new certificate and private key for an intermediate certificate authority
+func NewIntermediateCertificateAuthority(parentCert *x509.Certificate, parentKey crypto.Signer, config *CertConfig) (*x509.Certificate, crypto.Signer, error) {
```
config should be the first argument.

maybe the better API here is to have a single NewCertificateAuthority function that decides on self-signed vs signed-by-parent depending on a `nil` `parentCert`, but having a separate function is fine too.

robbiemcmichael replied:
Change made in this commit: https://github.com/kubernetes/kubernetes/pull/97266/commits/6f90532f2db098c52a531bb17e6db39d23a8db41

Since creating an intermediate CA is an uncommon use case (it's only used in the tests for the certificate chain verification) I think it's preferable not to force callers of `NewCertificateAuthority` (of which there are quite a few) to have to specify a `nil` certificate and key.

neolit123 reviewed (commented):
@robbiemcmichael 
thank you for the PR, i think the change is not necessarily bad, but i have concerns on how this plays with existing setups of kubeadm. have you though about what happens during upgrade on cert renew?

also unless i'm wrong the fact that the intermediate CA follows a leaf in your setup is not standardized, is it?

for the above and similar reasons, that is why we ask contributors to write a detailed proposal (KEP):
https://github.com/kubernetes/enhancements/tree/master/keps

/cc @randomvariable @fabriziopandini

neolit123 commented:
/cc @wallrj 
PTAL too

robbiemcmichael commented:
Thanks for the review @neolit123 

> thank you for the PR, i think the change is not necessarily bad, but i have concerns on how this plays with existing setups of kubeadm. have you though about what happens during upgrade on cert renew?

It's backwards compatible with the way that kubeadm generates certificates (i.e. leaf certificates signed directly by the root CA). I haven't changed any of this in this PR, this is just to support certificate chains in full external CA mode where every certificate is created externally. Certificate renewal will be the same as before.

> also unless i'm wrong the fact that the intermediate CA follows a leaf in your setup is not standardized, is it?

It's definitely an assumption already made by the `crypto/tls` library which Kubernetes already depends heavily on. I've provided some more details [here](https://github.com/kubernetes/kubernetes/pull/97266#discussion_r542988892).

> for the above and similar reasons, that is why we ask contributors to write a detailed proposal (KEP):
> https://github.com/kubernetes/enhancements/tree/master/keps

I thought KEPs were for much bigger changes than this. Although this PR contains around 300 lines of changes, most of the complexity is in the test setup. The changes made to the kubeadm logic can be summarised as follows:
- When loading certificates from files, don't throw away everything after the first certificate (as is done [here](https://github.com/kubernetes/kubernetes/blob/6f90532f2db098c52a531bb17e6db39d23a8db41/cmd/kubeadm/app/util/pkiutil/pki_helpers.go#L288-L290)) since these may be intermediate certificates which are required to verify that there is valid chain
- Instead of [`CheckSignatureFrom`](https://golang.org/pkg/crypto/x509/#Certificate.CheckSignatureFrom) we should use [`Verify`](https://golang.org/pkg/crypto/x509/#Certificate.Verify) as the latter automatically handles certificate chains

randomvariable commented:
I don't see any issues with backwards compatibility here. If the certificate is directly signed by the root with an empty intermediate certificate list, that should validate just fine.

randomvariable commented:
/ok-to-test

robbiemcmichael commented:
/retest

neolit123 commented on cmd/kubeadm/app/util/pkiutil/pki_helpers_test.go:
```diff
@@ -199,12 +199,27 @@ func TestWriteCert(t *testing.T) {
+	actual := WriteCertBundle(tmpdir, "foo", certs)
+	if actual != nil {
+		t.Errorf("failed WriteCertBundle with an error: %v", actual)
+	}
```
feels like this test should also verify that "foo" contains more than one cert?

neolit123 commented on cmd/kubeadm/app/util/pkiutil/pki_helpers_test.go:
```diff
@@ -804,3 +904,97 @@ func TestRemoveDuplicateAltNames(t *testing.T) {
+		Config: certutil.Config{CommonName: "Root CA 1"},
+	})
+	if err != nil {
+		t.Errorf("failed to create root CA cert and key with an error: %v", err)
```
t.Fatalf

robbiemcmichael replied:
Good catch, I copied that from some of the other tests. I've updated it for my tests and other tests where I think it should be `Fatalf`: https://github.com/kubernetes/kubernetes/pull/97266/commits/625491fe5ab296185448937464427ae1c3532840

neolit123 reviewed (commented):
/triage accepted
/kind feature

please keep the commits squashed to 2.
- implementation
- tests

neolit123 commented:
/priority backlog

neolit123 commented:
in the OP, under:
> Does this PR introduce a user-facing change?:

please add the following instead of NONE:
> kubeadm: add support .... <describe the feature with a couple of sentences>

this will add a release note for 1.21.

neolit123 commented on cmd/kubeadm/app/util/pkiutil/pki_helpers.go:
```diff
@@ -142,6 +158,26 @@ func WriteCert(pkiPath, name string, cert *x509.Certificate) error {
+func WriteCertBundle(pkiPath, name string, certs []*x509.Certificate) error {
+	for _, cert := range certs {
+		if cert == nil {
+			return errors.New("certificate cannot be nil when writing to file")
```
can we include an index `i` in this loop and modify the error to the following:

```
return errors.Errorf("found nil certificate at position %d when writing bundle to file", i)
```

robbiemcmichael replied:
Done in this commit: https://github.com/kubernetes/kubernetes/pull/97266/commits/945ca604f4d26f8805cc390191b29b7da5db761d

neolit123 commented on cmd/kubeadm/app/util/pkiutil/pki_helpers.go:
```diff
@@ -79,14 +80,29 @@ func NewCertificateAuthority(config *CertConfig) (*x509.Certificate, crypto.Sign
 }
 
+// NewIntermediateCertificateAuthority creates new certificate and private key for an intermediate certificate authority
+func NewIntermediateCertificateAuthority(parentCert *x509.Certificate, parentKey crypto.Signer, config *CertConfig) (*x509.Certificate, crypto.Signer, error) {
```
> Change made in this commit:

latest commit still has `config` as the last function argument.

should be:
```
NewIntermediateCertificateAuthority(config *CertConfig, parentCert *x509.Certificate, parentKey crypto.Signer)...
```

robbiemcmichael replied:
Sorry, the comment about the change being made was intended for the `isCA` bool being moved to be the last argument. I've edited the comments so that it makes sense now.

As for moving `config` to the last argument, I've gone with this ordering for consistency with th existing function [`NewCertAndKey`](https://github.com/kubernetes/kubernetes/blob/b9ce4ac212d150212485fa29d62a2fbd783a57b0/cmd/kubeadm/app/util/pkiutil/pki_helpers.go#L83). Let me know if you would prefer it to be consistent, different or whether I should update both.

neolit123 replied:
ok, let's leave the ordering as is, for consistency with NewCertAndKey.
the pkiutils package has no API guaranties yet, but we may have to review some of these semantics at some point.

neolit123 commented on cmd/kubeadm/app/util/pkiutil/pki_helpers.go:
```diff
@@ -556,32 +625,39 @@ func NewPrivateKey(keyType x509.PublicKeyAlgorithm) (crypto.Signer, error) {
 		return nil, errors.New("must specify a CommonName")
 	}
-	if len(cfg.Usages) == 0 {
+	if !isCA && len(cfg.Usages) == 0 {
```
i wonder if we should just remove the mandatory `len(cfg.Usages) > 0` here?
i believe currently we are passing either `x509.ExtKeyUsageClientAuth` or `x509.ExtKeyUsageServerAuth` to this function via config.

robbiemcmichael replied:
The config is passed through from `NewCertAndKey` so it's probably cleaner if I just perform this check there instead. I'll push a commit for this, let me know if you prefer that version otherwise I'm happy to remove it entirely.

neolit123 replied:
> The config is passed through from NewCertAndKey so it's probably cleaner if I just perform this check there instead.

agreed.

neolit123 commented on cmd/kubeadm/app/util/pkiutil/pki_helpers.go:
```diff
@@ -556,32 +625,39 @@ func NewPrivateKey(keyType x509.PublicKeyAlgorithm) (crypto.Signer, error) {
+		KeyUsage:              keyUsage,
+		ExtKeyUsage:           cfg.Usages,
+		BasicConstraintsValid: isCA,
+		IsCA:                  isCA,
```
https://golang.org/pkg/crypto/x509/

>     // BasicConstraintsValid indicates whether IsCA, MaxPathLen,
>    // and MaxPathLenZero are valid.

the documentation is a bit vague on what "valid" means, but should we assume `BasicConstraintsValid` should be always bound to `isCA`?

e.g. here it's not:
https://github.com/kubernetes/kubernetes/blob/dbfc3aa877b5beda2ecdfb38911c311b250a4973/pkg/controller/certificates/authority/policies.go#L65-L66

can we pass `true` always to it and an explicit value to `MaxPathLen` and `MaxPathLenZero=false`?

i'm assuming this is where it's checked:
https://github.com/golang/go/blob/master/src/crypto/x509/verify.go#L721-L726

robbiemcmichael replied:
I just did it this way to make sure I didn't make any changes to the existing code which signs leaf certificates. My interpretation of the docs is that it would probably work to set `BasicConstraintsValid` to `true`. We shouldn't need to explicitly set `MaxPathLen` and `MaxPathLenZero` since the default values of 0 and false respectively are "interpreted as `MaxPathLen` not being set" according the docs.

neolit123 replied:
my vote goes to set it always to `true` even for non-CA certificates.
`BasicConstraintsValid` seems like a Golang wrapper over the Basic Constraints extension, to conform checks in crypto/x509/verify.go and it doesn't seem like a breaking change.

robbiemcmichael commented:
Thanks for the review again @neolit123, I think I've made the changes or otherwise responded to all your comments but let me know if I've missed anything.

> please keep the commits squashed to 2.
> 
> implementation
> tests

I've just been using the extra commits to aid people in seeing the changes during the review process. I'll separate it into these two commits once you're happy that I've addressed all your comments.

neolit123 reviewed (commented):
@robbiemcmichael i went over the change again and it SGTM.
since @randomvariable is OK with this too, we can proceed with merging once you squash the commits.

thanks!
/approve

neolit123 commented:
/hold

robbiemcmichael committed:
kubeadm: support certificate chain validation

Fixes an issue where some kubeadm phases fail if a certificate file
contains a certificate chain with one or more intermediate CA
certificates. The validation algorithm has been changed from requiring
that a certificate was signed directly by the root CA to requiring that
there is a valid certificate chain back to the root CA.

robbiemcmichael committed:
kubeadm: tests for certificate chain validation

robbiemcmichael commented:
@neolit123 thanks again for the review, commits have been squashed and separated into implementation and tests as requested.

neolit123 reviewed (commented):
/lgtm
/hold cancel
